YoutubeExplode/Internal/YoutubeVideoStream.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace YoutubeExplode.Internal
+{
+    internal class YoutubeVideoStream : Stream
+    {
+        private readonly HttpClient _httpClient;
+
+        private const int MaxSegmentSize = 9_898_989;
+
+        private readonly string _url;
+
+        Stream _currentStream;
+
+        public YoutubeVideoStream(HttpClient httpClient, string url, long length)
+        {
+            _url = url;
+            _httpClient = httpClient;
+            Length = length;
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        public override long Length { get; }
+
+        public override long Position { get; set; }
+
+        public override void Flush() => throw new System.NotSupportedException();
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (Position >= Length)
+                return 0;
+
+            count = Math.Min(MaxSegmentSize - 1, count);
+            if(_currentStream == null)
+                _currentStream = await _httpClient.GetStreamAsync(_url, Position, Position + MaxSegmentSize).ConfigureAwait(false);
+            
+            var bytesRead = await _currentStream.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+            Position += bytesRead;
+            if (bytesRead == 0)
+            {
+                _currentStream.Dispose();
+                _currentStream = null;
+                bytesRead = await ReadAsync(buffer, offset, count, cancellationToken);
+            }
+            return bytesRead;
+        }
+        public override int Read(byte[] buffer, int offset, int count) =>
+            ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
+
+        public override long Seek(long offset, SeekOrigin origin)
```

Comment:
I'm not sure how Seeks work, I probably should throw instead of math'ing out what I can works with since user might get what's not expected

Reply:
[MemoryStream.Seek Implementation](https://referencesource.microsoft.com/#mscorlib/system/io/memorystream.cs,30f00da49c798fe1,references)